### DISCUSSION TITLE
Is `useSearchParams` memoized?

### DISCUSSION THREAD

**Discussion by BrianMcBrayer** [remix-run/react-router#8731] [Q&A]
Hello,

Thank you for react router!

I am curious if the result of `useSearchParams` is stable (memoized between renders)? In other words, if the search params haven't changed between renders, will this return the same object, so it can be used in a `useEffect` or `useCallback` dependency list?

Thanks!

-Brian

**Comment by BrianMcBrayer** [comment]
I... think(?) I have found the answer, and it is "no, it is not memoized". My example code:

```
const searchParams1 = useSearchParams();
const searchParams2 = useSearchParams();

console.log(searchParams1 === searchParams2);

useEffect(() => {
  console.log(`Different!!`);
}, [searchParams1]);
```
This logs `false` and `"Different!!"` every render.

**Reply by Noyabronok** [reply]
Dependency array values are compared with `Object.is` not `===`.  
But that's not the important part.  Hook stability can'be be compared across different hook calls.  You must compare the response from the same hook call on rerender.  Like this:

```
  // trigger and log rerendering
  const [renderCount, setRenderCount] = React.useState(0)
  setTimeout(() => setRenderCount((prev) => prev + 1), 1000)
  console.log(`Render count: ${renderCount}`)

  // single call to hook we're testing
  const [searchParams, setSearchParams] = useSearchParams()
  const paramRef = useRef()

  console.log('===', searchParams === paramRef.current)
  console.log('Object.is', Object.is(searchParams, paramRef.current))
  useEffect(() => {
    console.log(`Different!!`)
  }, [searchParams])

  // keep the last rendered value in the reference
  paramRef.current = searchParams
```

### Output:
```
Render count: 0
DevApp.tsx:81 === false
DevApp.tsx:82 Object.is false
DevApp.tsx:84 Different!!
DevApp.tsx:76 Render count: 1
DevApp.tsx:81 === true
DevApp.tsx:82 Object.is true
DevApp.tsx:76 Render count: 2
DevApp.tsx:81 === true
DevApp.tsx:82 Object.is true
DevApp.tsx:76 Render count: 3
DevApp.tsx:81 === true
DevApp.tsx:82 Object.is true
```

As you can see the `searchParams` response is stable
